File: src/services/transactionService.ts
import { httpClient } from './httpClient';
import { API_ENDPOINTS } from './apiConfig';
import { BulkPaymentBatch, BulkPaymentBatchItem, ChequeBookInventory, ChequeBookLeaf, ChequeClearingItem, Transaction } from '../../types';

interface TransactionApiModel {
  id: string;
  accountId?: string | null;
  type?: string | null;
  amount?: number;
  narration?: string | null;
  date?: string | null;
  tellerId?: string | null;
  status?: string | null;
  reference?: string | null;
}

interface BulkPaymentItemApiModel {
  id: string;
  accountId?: string | null;
  transactionType?: string | null;
  amount?: number;
  narration?: string | null;
  tellerId?: string | null;
  clientReference?: string | null;
  status?: string | null;
  postedTransactionId?: string | null;
  errorMessage?: string | null;
  processedAt?: string | null;
}

interface BulkPaymentBatchApiModel {
  id: string;
  batchReference?: string | null;
  status?: string | null;
  currency?: string | null;
  narration?: string | null;
  totalAmount?: number;
  processedAmount?: number;
  itemCount?: number;
  processedCount?: number;
  failedCount?: number;
  createdAt?: string | null;
  processedAt?: string | null;
  items?: BulkPaymentItemApiModel[] | null;
}

interface ChequeItemApiModel {
  id: string;
  accountId?: string | null;
  transactionType?: string | null;
  chequeNumber?: string | null;
  drawerName?: string | null;
  drawerAccountNumber?: string | null;
  presentingBankCode?: string | null;
  draweeBankCode?: string | null;
  clearingChannel?: string | null;
  bogRegulatoryClass?: string | null;
  isOtherBankCheque?: boolean | null;
  amount?: number;
  currency?: string | null;
  status?: string | null;
  holdDays?: number;
  lodgedAt?: string | null;
  clearingDate?: string | null;
  clearedAt?: string | null;
  postedTransactionId?: string | null;
  returnReason?: string | null;
  failureReason?: string | null;
  narration?: string | null;
}

interface ChequeBookLeafApiModel {
  id: string;
  serialNumber?: number;
  chequeNumber?: string | null;
  status?: string | null;
  accountId?: string | null;
  usedTransactionId?: string | null;
  usedAt?: string | null;
  cancelReason?: string | null;
}

interface ChequeBookInventoryApiModel {
  id: string;
  bookReference?: string | null;
  branchId?: string | null;
  seriesPrefix?: string | null;
  startSerialNumber?: number;
  endSerialNumber?: number;
  leafCount?: number;
  availableLeafCount?: number;
  usedLeafCount?: number;
  cancelledLeafCount?: number;
  status?: string | null;
  accountId?: string | null;
  customerId?: string | null;
  issuedAt?: string | null;
  issuedBy?: string | null;
  remarks?: string | null;
  createdAt?: string | null;
  leaves?: ChequeBookLeafApiModel[] | null;
}

type NormalizedTransactionType = Transaction['type'];
type BulkStatus = BulkPaymentBatch['status'];
type ChequeStatus = ChequeClearingItem['status'];
type ChequeBookStatus = ChequeBookInventory['status'];

const normalizeTransactionType = (value?: string | null): NormalizedTransactionType => {
  const normalized = (value || 'DEPOSIT').trim().toUpperCase();
  if (normalized === 'WITHDRAWAL') return 'WITHDRAWAL';
  if (normalized === 'TRANSFER') return 'TRANSFER';
  if (normalized === 'LOAN_REPAYMENT') return 'LOAN_REPAYMENT';
  return 'DEPOSIT';
};

const normalizeTransactionStatus = (value?: string | null): Transaction['status'] => {
  const normalized = (value || 'POSTED').trim().toUpperCase();
  if (normalized === 'PENDING' || normalized === 'PENDING_APPROVAL') return 'PENDING';
  if (normalized === 'REJECTED' || normalized === 'FAILED') return 'REJECTED';
  return 'POSTED';
};

const normalizeBulkStatus = (value?: string | null): BulkStatus => {
  const normalized = (value || 'PROCESSING').trim().toUpperCase();
  if (normalized === 'COMPLETED') return 'COMPLETED';
  if (normalized === 'FAILED') return 'FAILED';
  if (normalized === 'PARTIAL') return 'PARTIAL';
  if (normalized === 'POSTED') return 'POSTED';
  if (normalized === 'PENDING') return 'PENDING';
  return 'PROCESSING';
};

const normalizeChequeStatus = (value?: string | null): ChequeStatus => {
  const normalized = (value || 'LODGED').trim().toUpperCase();
  if (normalized === 'PENDING_CLEARING') return 'PENDING_CLEARING';
  if (normalized === 'CLEARED') return 'CLEARED';
  if (normalized === 'PAID') return 'PAID';
  if (normalized === 'RETURNED') return 'RETURNED';
  if (normalized === 'FAILED') return 'FAILED';
  return 'LODGED';
};

const normalizeChequeBookStatus = (value?: string | null): ChequeBookStatus => {
  const normalized = (value || 'IN_STOCK').trim().toUpperCase();
  if (normalized === 'ISSUED') return 'ISSUED';
  if (normalized === 'ACTIVE') return 'ACTIVE';
  if (normalized === 'EXHAUSTED') return 'EXHAUSTED';
  return 'IN_STOCK';
};

const mapTransaction = (transaction: TransactionApiModel): Transaction => ({
  id: transaction.id,
  accountId: transaction.accountId || '',
  type: normalizeTransactionType(transaction.type),
  amount: Number(transaction.amount || 0),
  narration: transaction.narration || '',
  date: transaction.date || new Date().toISOString(),
  tellerId: transaction.tellerId || '',
  status: normalizeTransactionStatus(transaction.status),
  reference: transaction.reference || '',
});

const mapBulkItem = (item: BulkPaymentItemApiModel): BulkPaymentBatchItem => ({
  id: item.id,
  accountId: item.accountId || '',
  transactionType: normalizeTransactionType(item.transactionType),
  amount: Number(item.amount || 0),
  narration: item.narration || '',
  tellerId: item.tellerId || '',
  clientReference: item.clientReference || '',
  status: normalizeBulkStatus(item.status),
  postedTransactionId: item.postedTransactionId || '',
  errorMessage: item.errorMessage || '',
  processedAt: item.processedAt || undefined,
});

const mapBulkBatch = (batch: BulkPaymentBatchApiModel): BulkPaymentBatch => ({
  id: batch.id,
  batchReference: batch.batchReference || '',
  status: normalizeBulkStatus(batch.status),
  currency: batch.currency || 'GHS',
  narration: batch.narration || '',
  totalAmount: Number(batch.totalAmount || 0),
  processedAmount: Number(batch.processedAmount || 0),
  itemCount: Number(batch.itemCount || 0),
  processedCount: Number(batch.processedCount || 0),
  failedCount: Number(batch.failedCount || 0),
  createdAt: batch.createdAt || new Date().toISOString(),
  processedAt: batch.processedAt || undefined,
  items: (batch.items || []).map(mapBulkItem),
});

const mapChequeItem = (item: ChequeItemApiModel): ChequeClearingItem => ({
  id: item.id,
  accountId: item.accountId || '',
  transactionType: normalizeTransactionType(item.transactionType) as ChequeClearingItem['transactionType'],
  chequeNumber: item.chequeNumber || '',
  drawerName: item.drawerName || '',
  drawerAccountNumber: item.drawerAccountNumber || '',
  presentingBankCode: item.presentingBankCode || '',
  draweeBankCode: item.draweeBankCode || '',
  clearingChannel: item.clearingChannel || '',
  bogRegulatoryClass: item.bogRegulatoryClass || '',
  isOtherBankCheque: Boolean(item.isOtherBankCheque),
  amount: Number(item.amount || 0),
  currency: item.currency || 'GHS',
  status: normalizeChequeStatus(item.status),
  holdDays: Number(item.holdDays || 0),
  lodgedAt: item.lodgedAt || new Date().toISOString(),
  clearingDate: item.clearingDate || '',
  clearedAt: item.clearedAt || undefined,
  postedTransactionId: item.postedTransactionId || '',
  returnReason: item.returnReason || '',
  failureReason: item.failureReason || '',
  narration: item.narration || '',
});

const mapChequeBookLeaf = (leaf: ChequeBookLeafApiModel): ChequeBookLeaf => ({
  id: leaf.id,
  serialNumber: Number(leaf.serialNumber || 0),
  chequeNumber: leaf.chequeNumber || '',
  status: ((leaf.status || 'AVAILABLE').trim().toUpperCase() as ChequeBookLeaf['status']),
  accountId: leaf.accountId || '',
  usedTransactionId: leaf.usedTransactionId || '',
  usedAt: leaf.usedAt || undefined,
  cancelReason: leaf.cancelReason || '',
});

const mapChequeBook = (book: ChequeBookInventoryApiModel): ChequeBookInventory => ({
  id: book.id,
  bookReference: book.bookReference || '',
  branchId: book.branchId || '',
  seriesPrefix: book.seriesPrefix || '',
  startSerialNumber: Number(book.startSerialNumber || 0),
  endSerialNumber: Number(book.endSerialNumber || 0),
  leafCount: Number(book.leafCount || 0),
  availableLeafCount: Number(book.availableLeafCount || 0),
  usedLeafCount: Number(book.usedLeafCount || 0),
  cancelledLeafCount: Number(book.cancelledLeafCount || 0),
  status: normalizeChequeBookStatus(book.status),
  accountId: book.accountId || '',
  customerId: book.customerId || '',
  issuedAt: book.issuedAt || undefined,
  issuedBy: book.issuedBy || '',
  remarks: book.remarks || '',
  createdAt: book.createdAt || new Date().toISOString(),
  leaves: (book.leaves || []).map(mapChequeBookLeaf),
});

class TransactionService {
  async getTransactions(): Promise<Transaction[]> {
    const transactions = await httpClient.get<TransactionApiModel[]>(API_ENDPOINTS.transactions.list);
    return transactions.map(mapTransaction);
  }

  async createTransaction(data: {
    accountId: string;
    type: Transaction['type'];
    amount: number;
    narration?: string;
    tellerId: string;
    clientReference?: string;
  }): Promise<Transaction> {
    const created = await httpClient.post<TransactionApiModel>(API_ENDPOINTS.transactions.list, data);
    return mapTransaction(created);
  }

  async getBulkPaymentBatches(): Promise<BulkPaymentBatch[]> {
    const batches = await httpClient.get<BulkPaymentBatchApiModel[]>(API_ENDPOINTS.payments.bulk);
    return batches.map(mapBulkBatch);
  }

  async createBulkPaymentBatch(data: {
    currency: string;
    narration?: string;
    submittedBy?: string;
    items: Array<{
      accountId: string;
      transactionType: Transaction['type'];
      amount: number;
      narration?: string;
      tellerId?: string;
      clientReference?: string;
    }>;
  }): Promise<BulkPaymentBatch> {
    const created = await httpClient.post<BulkPaymentBatchApiModel>(API_ENDPOINTS.payments.bulk, data);
    return mapBulkBatch(created);
  }

  async getChequeItems(): Promise<ChequeClearingItem[]> {
    const cheques = await httpClient.get<ChequeItemApiModel[]>(API_ENDPOINTS.payments.cheques);
    return cheques.map(mapChequeItem);
  }

  async lodgeChequeDeposit(data: {
    accountId: string;
    chequeNumber: string;
    amount: number;
    currency: string;
    drawerName?: string;
    drawerAccountNumber?: string;
    presentingBankCode: string;
    draweeBankCode: string;
    isOtherBankCheque: boolean;
    clearingChannel: string;
    bogRegulatoryClass: string;
    tellerId?: string;
    narration?: string;
  }): Promise<ChequeClearingItem> {
    const created = await httpClient.post<ChequeItemApiModel>(API_ENDPOINTS.payments.chequeDeposits, data);
    return mapChequeItem(created);
  }

  async processChequeWithdrawal(data: {
    accountId: string;
    chequeNumber: string;
    amount: number;
    currency: string;
    tellerId: string;
    narration?: string;
  }): Promise<ChequeClearingItem> {
    const created = await httpClient.post<ChequeItemApiModel>(API_ENDPOINTS.payments.chequeWithdrawals, data);
    return mapChequeItem(created);
  }

  async returnCheque(itemId: string, reason: string): Promise<ChequeClearingItem> {
    const updated = await httpClient.post<ChequeItemApiModel>(API_ENDPOINTS.payments.chequeReturn(itemId), { reason });
    return mapChequeItem(updated);
  }

  async getChequeBooks(accountId?: string): Promise<ChequeBookInventory[]> {
    const endpoint = accountId
      ? `${API_ENDPOINTS.payments.chequeBooks}?accountId=${encodeURIComponent(accountId)}`
      : API_ENDPOINTS.payments.chequeBooks;
    const books = await httpClient.get<ChequeBookInventoryApiModel[]>(endpoint);
    return books.map(mapChequeBook);
  }

  async createChequeBookStock(data: {
    branchId: string;
    seriesPrefix: string;
    startSerialNumber: number;
    leafCount: number;
    remarks?: string;
  }): Promise<ChequeBookInventory> {
    const created = await httpClient.post<ChequeBookInventoryApiModel>(API_ENDPOINTS.payments.chequeBookStock, data);
    return mapChequeBook(created);
  }

  async issueChequeBook(bookId: string, data: { accountId: string; issuedBy?: string; remarks?: string }): Promise<ChequeBookInventory> {
    const updated = await httpClient.post<ChequeBookInventoryApiModel>(API_ENDPOINTS.payments.chequeBookIssue(bookId), data);
    return mapChequeBook(updated);
  }
}

export const transactionService = new TransactionService();

File: CoreBankerWeb/CoreBanker/Services/TransactionService.cs
using System.Globalization;

namespace CoreBanker.Services
{
    public class TransactionService : ApiClientBase
    {
        public TransactionService(HttpClient httpClient, CoreBanker.State.AppState appState) : base(httpClient, appState) { }

        public async Task<List<TransactionDto>> GetTransactionsAsync(CancellationToken cancellationToken = default)
        {
            var result = await GetAsync<List<TransactionApiModel>>("/api/transactions", cancellationToken);
            return (result ?? new List<TransactionApiModel>()).ConvertAll(MapTransaction);
        }

        public async Task<List<BulkPaymentBatchDto>> GetBulkPaymentBatchesAsync(CancellationToken cancellationToken = default)
        {
            var result = await GetAsync<List<BulkPaymentBatchApiModel>>("/api/payments/bulk", cancellationToken);
            return (result ?? new List<BulkPaymentBatchApiModel>()).ConvertAll(MapBulkBatch);
        }

        public async Task<BulkPaymentBatchDto?> CreateBulkPaymentBatchAsync(BulkPaymentBatchRequest request, CancellationToken cancellationToken = default)
        {
            var response = await PostAsync<BulkPaymentBatchRequest, BulkPaymentBatchApiModel>("/api/payments/bulk", request, cancellationToken);
            return response is null ? null : MapBulkBatch(response);
        }

        public async Task<List<ChequeClearingItemDto>> GetChequeItemsAsync(CancellationToken cancellationToken = default)
        {
            var result = await GetAsync<List<ChequeClearingItemApiModel>>("/api/payments/cheques", cancellationToken);
            return (result ?? new List<ChequeClearingItemApiModel>()).ConvertAll(MapChequeItem);
        }

        public async Task<ChequeClearingItemDto?> LodgeChequeDepositAsync(ChequeDepositRequest request, CancellationToken cancellationToken = default)
        {
            var response = await PostAsync<ChequeDepositRequest, ChequeClearingItemApiModel>("/api/payments/cheques/deposits", request, cancellationToken);
            return response is null ? null : MapChequeItem(response);
        }

        public async Task<ChequeClearingItemDto?> ProcessChequeWithdrawalAsync(ChequeWithdrawalRequest request, CancellationToken cancellationToken = default)
        {
            var response = await PostAsync<ChequeWithdrawalRequest, ChequeClearingItemApiModel>("/api/payments/cheques/withdrawals", request, cancellationToken);
            return response is null ? null : MapChequeItem(response);
        }

        public async Task<ChequeClearingItemDto?> ReturnChequeAsync(string itemId, string reason, CancellationToken cancellationToken = default)
        {
            var response = await PostAsync<ReturnChequeRequest, ChequeClearingItemApiModel>($"/api/payments/cheques/{itemId}/return", new ReturnChequeRequest { Reason = reason }, cancellationToken);
            return response is null ? null : MapChequeItem(response);
        }

        public async Task<List<ChequeBookInventoryDto>> GetChequeBooksAsync(string? accountId = null, CancellationToken cancellationToken = default)
        {
            var path = string.IsNullOrWhiteSpace(accountId)
                ? "/api/payments/cheque-books"
                : $"/api/payments/cheque-books?accountId={Uri.EscapeDataString(accountId)}";
            var result = await GetAsync<List<ChequeBookInventoryApiModel>>(path, cancellationToken);
            return (result ?? new List<ChequeBookInventoryApiModel>()).ConvertAll(MapChequeBook);
        }

        public async Task<ChequeBookInventoryDto?> CreateChequeBookStockAsync(CreateChequeBookStockRequest request, CancellationToken cancellationToken = default)
        {
            var response = await PostAsync<CreateChequeBookStockRequest, ChequeBookInventoryApiModel>("/api/payments/cheque-books/stock", request, cancellationToken);
            return response is null ? null : MapChequeBook(response);
        }

        public async Task<ChequeBookInventoryDto?> IssueChequeBookAsync(string bookId, IssueChequeBookRequest request, CancellationToken cancellationToken = default)
        {
            var response = await PostAsync<IssueChequeBookRequest, ChequeBookInventoryApiModel>($"/api/payments/cheque-books/{bookId}/issue", request, cancellationToken);
            return response is null ? null : MapChequeBook(response);
        }

        public async Task<ChequeBookInventoryDto?> CancelChequeLeafAsync(string leafId, string reason, CancellationToken cancellationToken = default)
        {
            var response = await PostAsync<CancelChequeLeafRequest, ChequeBookInventoryApiModel>($"/api/payments/cheque-books/leaves/{leafId}/cancel", new CancelChequeLeafRequest { Reason = reason }, cancellationToken);
            return response is null ? null : MapChequeBook(response);
        }

        private static TransactionDto MapTransaction(TransactionApiModel txn)
        {
            return new TransactionDto
            {
                Id = txn.Id ?? string.Empty,
                AccountId = txn.AccountId ?? string.Empty,
                Type = NormalizeType(txn.Type),
                Amount = txn.Amount ?? 0m,
                Date = ParseDate(txn.Date),
                Status = NormalizeStatus(txn.Status),
                Narration = txn.Narration ?? string.Empty,
                Reference = txn.Reference ?? string.Empty,
                TellerId = txn.TellerId ?? string.Empty
            };
        }

        private static BulkPaymentBatchDto MapBulkBatch(BulkPaymentBatchApiModel batch)
        {
            return new BulkPaymentBatchDto
            {
                Id = batch.Id ?? string.Empty,
                BatchReference = batch.BatchReference ?? string.Empty,
                Status = NormalizeStatus(batch.Status),
                Currency = batch.Currency ?? "GHS",
                Narration = batch.Narration ?? string.Empty,
                TotalAmount = batch.TotalAmount ?? 0m,
                ProcessedAmount = batch.ProcessedAmount ?? 0m,
                ItemCount = batch.ItemCount ?? 0,
                ProcessedCount = batch.ProcessedCount ?? 0,
                FailedCount = batch.FailedCount ?? 0,
                CreatedAt = ParseDate(batch.CreatedAt),
                ProcessedAt = string.IsNullOrWhiteSpace(batch.ProcessedAt) ? null : ParseDate(batch.ProcessedAt),
                Items = (batch.Items ?? new List<BulkPaymentItemApiModel>()).ConvertAll(MapBulkItem)
            };
        }

        private static BulkPaymentItemDto MapBulkItem(BulkPaymentItemApiModel item)
        {
            return new BulkPaymentItemDto
            {
                Id = item.Id ?? string.Empty,
                AccountId = item.AccountId ?? string.Empty,
                TransactionType = NormalizeType(item.TransactionType),
                Amount = item.Amount ?? 0m,
                Narration = item.Narration ?? string.Empty,
                TellerId = item.TellerId ?? string.Empty,
                ClientReference = item.ClientReference ?? string.Empty,
                Status = NormalizeStatus(item.Status),
                PostedTransactionId = item.PostedTransactionId ?? string.Empty,
                ErrorMessage = item.ErrorMessage ?? string.Empty,
                ProcessedAt = string.IsNullOrWhiteSpace(item.ProcessedAt) ? null : ParseDate(item.ProcessedAt)
            };
        }

        private static ChequeClearingItemDto MapChequeItem(ChequeClearingItemApiModel item)
        {
            return new ChequeClearingItemDto
            {
                Id = item.Id ?? string.Empty,
                AccountId = item.AccountId ?? string.Empty,
                TransactionType = NormalizeType(item.TransactionType),
                ChequeNumber = item.ChequeNumber ?? string.Empty,
                DrawerName = item.DrawerName ?? string.Empty,
                DrawerAccountNumber = item.DrawerAccountNumber ?? string.Empty,
                PresentingBankCode = item.PresentingBankCode ?? string.Empty,
                DraweeBankCode = item.DraweeBankCode ?? string.Empty,
                ClearingChannel = item.ClearingChannel ?? string.Empty,
                BogRegulatoryClass = item.BogRegulatoryClass ?? string.Empty,
                IsOtherBankCheque = item.IsOtherBankCheque ?? false,
                Amount = item.Amount ?? 0m,
                Currency = item.Currency ?? "GHS",
                Status = NormalizeStatus(item.Status),
                HoldDays = item.HoldDays ?? 0,
                LodgedAt = ParseDate(item.LodgedAt),
                ClearingDate = item.ClearingDate ?? string.Empty,
                ClearedAt = string.IsNullOrWhiteSpace(item.ClearedAt) ? null : ParseDate(item.ClearedAt),
                PostedTransactionId = item.PostedTransactionId ?? string.Empty,
                ReturnReason = item.ReturnReason ?? string.Empty,
                FailureReason = item.FailureReason ?? string.Empty,
                Narration = item.Narration ?? string.Empty
            };
        }

        private static ChequeBookInventoryDto MapChequeBook(ChequeBookInventoryApiModel book)
        {
            return new ChequeBookInventoryDto
            {
                Id = book.Id ?? string.Empty,
                BookReference = book.BookReference ?? string.Empty,
                BranchId = book.BranchId ?? string.Empty,
                SeriesPrefix = book.SeriesPrefix ?? string.Empty,
                StartSerialNumber = book.StartSerialNumber ?? 0,
                EndSerialNumber = book.EndSerialNumber ?? 0,
                LeafCount = book.LeafCount ?? 0,
                AvailableLeafCount = book.AvailableLeafCount ?? 0,
                UsedLeafCount = book.UsedLeafCount ?? 0,
                CancelledLeafCount = book.CancelledLeafCount ?? 0,
                Status = book.Status ?? string.Empty,
                AccountId = book.AccountId ?? string.Empty,
                CustomerId = book.CustomerId ?? string.Empty,
                IssuedAt = string.IsNullOrWhiteSpace(book.IssuedAt) ? null : ParseDate(book.IssuedAt),
                IssuedBy = book.IssuedBy ?? string.Empty,
                Remarks = book.Remarks ?? string.Empty,
                CreatedAt = ParseDate(book.CreatedAt),
                Leaves = (book.Leaves ?? new List<ChequeBookLeafApiModel>()).ConvertAll(MapChequeBookLeaf)
            };
        }

        private static ChequeBookLeafDto MapChequeBookLeaf(ChequeBookLeafApiModel leaf)
        {
            return new ChequeBookLeafDto
            {
                Id = leaf.Id ?? string.Empty,
                SerialNumber = leaf.SerialNumber ?? 0,
                ChequeNumber = leaf.ChequeNumber ?? string.Empty,
                Status = leaf.Status ?? string.Empty,
                AccountId = leaf.AccountId ?? string.Empty,
                UsedTransactionId = leaf.UsedTransactionId ?? string.Empty,
                UsedAt = string.IsNullOrWhiteSpace(leaf.UsedAt) ? null : ParseDate(leaf.UsedAt),
                CancelReason = leaf.CancelReason ?? string.Empty
            };
        }

        private static DateTime ParseDate(string? value)
        {
            return DateTime.TryParse(value, CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal, out var parsed)
                ? parsed
                : DateTime.UtcNow;
        }

        private static string NormalizeType(string? value)
        {
            var normalized = (value ?? "DEPOSIT").Trim().ToUpperInvariant();
            return normalized switch
            {
                "WITHDRAWAL" => "WITHDRAWAL",
                "TRANSFER" => "TRANSFER",
                "LOAN_REPAYMENT" => "LOAN_REPAYMENT",
                "DEPOSIT" => "DEPOSIT",
                _ => "DEPOSIT"
            };
        }

        private static string NormalizeStatus(string? value)
        {
            var normalized = (value ?? "POSTED").Trim().ToUpperInvariant();
            return normalized switch
            {
                "PENDING" => "PENDING",
                "FAILED" => "FAILED",
                "PROCESSING" => "PROCESSING",
                "COMPLETED" => "COMPLETED",
                "PARTIAL" => "PARTIAL",
                "LODGED" => "LODGED",
                "CLEARED" => "CLEARED",
                "PAID" => "PAID",
                "RETURNED" => "RETURNED",
                "PENDING_CLEARING" => "PENDING_CLEARING",
                _ => "POSTED"
            };
        }

        private sealed class TransactionApiModel
        {
            public string? Id { get; set; }
            public string? AccountId { get; set; }
            public string? Type { get; set; }
            public decimal? Amount { get; set; }
            public string? Date { get; set; }
            public string? Status { get; set; }
            public string? Narration { get; set; }
            public string? Reference { get; set; }
            public string? TellerId { get; set; }
        }

        private sealed class BulkPaymentBatchApiModel
        {
            public string? Id { get; set; }
            public string? BatchReference { get; set; }
            public string? Status { get; set; }
            public string? Currency { get; set; }
            public string? Narration { get; set; }
            public decimal? TotalAmount { get; set; }
            public decimal? ProcessedAmount { get; set; }
            public int? ItemCount { get; set; }
            public int? ProcessedCount { get; set; }
            public int? FailedCount { get; set; }
            public string? CreatedAt { get; set; }
            public string? ProcessedAt { get; set; }
            public List<BulkPaymentItemApiModel>? Items { get; set; }
        }

        private sealed class BulkPaymentItemApiModel
        {
            public string? Id { get; set; }
            public string? AccountId { get; set; }
            public string? TransactionType { get; set; }
            public decimal? Amount { get; set; }
            public string? Narration { get; set; }
            public string? TellerId { get; set; }
            public string? ClientReference { get; set; }
            public string? Status { get; set; }
            public string? PostedTransactionId { get; set; }
            public string? ErrorMessage { get; set; }
            public string? ProcessedAt { get; set; }
        }

        private sealed class ChequeClearingItemApiModel
        {
            public string? Id { get; set; }
            public string? AccountId { get; set; }
            public string? TransactionType { get; set; }
            public string? ChequeNumber { get; set; }
            public string? DrawerName { get; set; }
            public string? DrawerAccountNumber { get; set; }
            public string? PresentingBankCode { get; set; }
            public string? DraweeBankCode { get; set; }
            public string? ClearingChannel { get; set; }
            public string? BogRegulatoryClass { get; set; }
            public bool? IsOtherBankCheque { get; set; }
            public decimal? Amount { get; set; }
            public string? Currency { get; set; }
            public string? Status { get; set; }
            public int? HoldDays { get; set; }
            public string? LodgedAt { get; set; }
            public string? ClearingDate { get; set; }
            public string? ClearedAt { get; set; }
            public string? PostedTransactionId { get; set; }
            public string? ReturnReason { get; set; }
            public string? FailureReason { get; set; }
            public string? Narration { get; set; }
        }

        private sealed class ChequeBookInventoryApiModel
        {
            public string? Id { get; set; }
            public string? BookReference { get; set; }
            public string? BranchId { get; set; }
            public string? SeriesPrefix { get; set; }
            public long? StartSerialNumber { get; set; }
            public long? EndSerialNumber { get; set; }
            public int? LeafCount { get; set; }
            public int? AvailableLeafCount { get; set; }
            public int? UsedLeafCount { get; set; }
            public int? CancelledLeafCount { get; set; }
            public string? Status { get; set; }
            public string? AccountId { get; set; }
            public string? CustomerId { get; set; }
            public string? IssuedAt { get; set; }
            public string? IssuedBy { get; set; }
            public string? Remarks { get; set; }
            public string? CreatedAt { get; set; }
            public List<ChequeBookLeafApiModel>? Leaves { get; set; }
        }

        private sealed class ChequeBookLeafApiModel
        {
            public string? Id { get; set; }
            public long? SerialNumber { get; set; }
            public string? ChequeNumber { get; set; }
            public string? Status { get; set; }
            public string? AccountId { get; set; }
            public string? UsedTransactionId { get; set; }
            public string? UsedAt { get; set; }
            public string? CancelReason { get; set; }
        }
    }

    public class TransactionDto
    {
        public string Id { get; set; } = string.Empty;
        public string AccountId { get; set; } = string.Empty;
        public string Type { get; set; } = string.Empty;
        public decimal Amount { get; set; }
        public DateTime Date { get; set; }
        public string Status { get; set; } = "POSTED";
        public string Narration { get; set; } = string.Empty;
        public string Reference { get; set; } = string.Empty;
        public string TellerId { get; set; } = string.Empty;
    }

    public class BulkPaymentBatchRequest
    {
        public string Currency { get; set; } = "GHS";
        public string? Narration { get; set; }
        public string? SubmittedBy { get; set; }
        public List<BulkPaymentItemRequest> Items { get; set; } = new();
    }

    public class BulkPaymentItemRequest
    {
        public string AccountId { get; set; } = string.Empty;
        public string TransactionType { get; set; } = "DEPOSIT";
        public decimal Amount { get; set; }
        public string? Narration { get; set; }
        public string? TellerId { get; set; }
        public string? ClientReference { get; set; }
    }

    public class BulkPaymentBatchDto
    {
        public string Id { get; set; } = string.Empty;
        public string BatchReference { get; set; } = string.Empty;
        public string Status { get; set; } = string.Empty;
        public string Currency { get; set; } = "GHS";
        public string Narration { get; set; } = string.Empty;
        public decimal TotalAmount { get; set; }
        public decimal ProcessedAmount { get; set; }
        public int ItemCount { get; set; }
        public int ProcessedCount { get; set; }
        public int FailedCount { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime? ProcessedAt { get; set; }
        public List<BulkPaymentItemDto> Items { get; set; } = new();
    }

    public class BulkPaymentItemDto
    {
        public string Id { get; set; } = string.Empty;
        public string AccountId { get; set; } = string.Empty;
        public string TransactionType { get; set; } = string.Empty;
        public decimal Amount { get; set; }
        public string Narration { get; set; } = string.Empty;
        public string TellerId { get; set; } = string.Empty;
        public string ClientReference { get; set; } = string.Empty;
        public string Status { get; set; } = string.Empty;
        public string PostedTransactionId { get; set; } = string.Empty;
        public string ErrorMessage { get; set; } = string.Empty;
        public DateTime? ProcessedAt { get; set; }
    }

    public class ChequeDepositRequest
    {
        public string AccountId { get; set; } = string.Empty;
        public string ChequeNumber { get; set; } = string.Empty;
        public decimal Amount { get; set; }
        public string Currency { get; set; } = "GHS";
        public string? DrawerName { get; set; }
        public string? DrawerAccountNumber { get; set; }
        public string PresentingBankCode { get; set; } = string.Empty;
        public string DraweeBankCode { get; set; } = string.Empty;
        public bool IsOtherBankCheque { get; set; } = true;
        public string ClearingChannel { get; set; } = "GHIPSS";
        public string BogRegulatoryClass { get; set; } = "LOCAL";
        public string? TellerId { get; set; }
        public string? Narration { get; set; }
    }

    public class ChequeWithdrawalRequest
    {
        public string AccountId { get; set; } = string.Empty;
        public string ChequeNumber { get; set; } = string.Empty;
        public decimal Amount { get; set; }
        public string Currency { get; set; } = "GHS";
        public string TellerId { get; set; } = string.Empty;
        public string? Narration { get; set; }
    }

    public class ReturnChequeRequest
    {
        public string Reason { get; set; } = string.Empty;
    }

    public class ChequeClearingItemDto
    {
        public string Id { get; set; } = string.Empty;
        public string AccountId { get; set; } = string.Empty;
        public string TransactionType { get; set; } = string.Empty;
        public string ChequeNumber { get; set; } = string.Empty;
        public string DrawerName { get; set; } = string.Empty;
        public string DrawerAccountNumber { get; set; } = string.Empty;
        public string PresentingBankCode { get; set; } = string.Empty;
        public string DraweeBankCode { get; set; } = string.Empty;
        public string ClearingChannel { get; set; } = string.Empty;
        public string BogRegulatoryClass { get; set; } = string.Empty;
        public bool IsOtherBankCheque { get; set; }
        public decimal Amount { get; set; }
        public string Currency { get; set; } = "GHS";
        public string Status { get; set; } = string.Empty;
        public int HoldDays { get; set; }
        public DateTime LodgedAt { get; set; }
        public string ClearingDate { get; set; } = string.Empty;
        public DateTime? ClearedAt { get; set; }
        public string PostedTransactionId { get; set; } = string.Empty;
        public string ReturnReason { get; set; } = string.Empty;
        public string FailureReason { get; set; } = string.Empty;
        public string Narration { get; set; } = string.Empty;
    }

    public class CreateChequeBookStockRequest
    {
        public string BranchId { get; set; } = string.Empty;
        public string SeriesPrefix { get; set; } = string.Empty;
        public long StartSerialNumber { get; set; }
        public int LeafCount { get; set; }
        public string? Remarks { get; set; }
    }

    public class IssueChequeBookRequest
    {
        public string AccountId { get; set; } = string.Empty;
        public string? IssuedBy { get; set; }
        public string? Remarks { get; set; }
    }

    public class CancelChequeLeafRequest
    {
        public string Reason { get; set; } = string.Empty;
    }

    public class ChequeBookInventoryDto
    {
        public string Id { get; set; } = string.Empty;
        public string BookReference { get; set; } = string.Empty;
        public string BranchId { get; set; } = string.Empty;
        public string SeriesPrefix { get; set; } = string.Empty;
        public long StartSerialNumber { get; set; }
        public long EndSerialNumber { get; set; }
        public int LeafCount { get; set; }
        public int AvailableLeafCount { get; set; }
        public int UsedLeafCount { get; set; }
        public int CancelledLeafCount { get; set; }
        public string Status { get; set; } = string.Empty;
        public string AccountId { get; set; } = string.Empty;
        public string CustomerId { get; set; } = string.Empty;
        public DateTime? IssuedAt { get; set; }
        public string IssuedBy { get; set; } = string.Empty;
        public string Remarks { get; set; } = string.Empty;
        public DateTime CreatedAt { get; set; }
        public List<ChequeBookLeafDto> Leaves { get; set; } = new();
    }

    public class ChequeBookLeafDto
    {
        public string Id { get; set; } = string.Empty;
        public long SerialNumber { get; set; }
        public string ChequeNumber { get; set; } = string.Empty;
        public string Status { get; set; } = string.Empty;
        public string AccountId { get; set; } = string.Empty;
        public string UsedTransactionId { get; set; } = string.Empty;
        public DateTime? UsedAt { get; set; }
        public string CancelReason { get; set; } = string.Empty;
    }
}
